File: submit_application.py
#!/usr/bin/env python3
"""
Rental Application Submission Handler
Receives form data and stores it in JSON files
"""

from http.server import HTTPServer, BaseHTTPRequestHandler
import json
import os
from datetime import datetime
from urllib.parse import parse_qs
import cgi

# Configuration
SUBMISSIONS_DIR = "applications"
PORT = 8001

class ApplicationHandler(BaseHTTPRequestHandler):
    """Handle rental application submissions"""
    
    def do_OPTIONS(self):
        """Handle CORS preflight requests"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
    
    def do_POST(self):
        """Handle POST request with application data"""
        if self.path == '/submit':
            try:
                # Get content length
                content_length = int(self.headers.get('Content-Length', 0))
                
                # Read and parse JSON data
                post_data = self.rfile.read(content_length)
                application_data = json.loads(post_data.decode('utf-8'))
                
                # Add metadata
                application_data['submittedAt'] = datetime.now().isoformat()
                application_data['submittedFrom'] = self.client_address[0]
                
                # Create submissions directory if it doesn't exist
                if not os.path.exists(SUBMISSIONS_DIR):
                    os.makedirs(SUBMISSIONS_DIR)
                
                # Generate unique filename with timestamp
                timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
                applicant_name = f"{application_data.get('firstName', 'Unknown')}_{application_data.get('lastName', 'Unknown')}"
                applicant_name = applicant_name.replace(' ', '_').replace('/', '_')
                filename = f"{timestamp}_{applicant_name}.json"
                filepath = os.path.join(SUBMISSIONS_DIR, filename)
                
                # Save application data to JSON file
                with open(filepath, 'w') as f:
                    json.dump(application_data, f, indent=2)
                
                # Log submission
                print(f"✅ Application received: {applicant_name} - Saved to {filename}")
                
                # Send success response
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                
                response = {
                    'success': True,
                    'message': 'Application submitted successfully',
                    'filename': filename,
                    'timestamp': application_data['submittedAt']
                }
                self.wfile.write(json.dumps(response).encode('utf-8'))
                
            except json.JSONDecodeError as e:
                # JSON parsing error
                self.send_error(400, f"Invalid JSON: {str(e)}")
                print(f"❌ JSON Error: {e}")
                
            except Exception as e:
                # General error
                self.send_error(500, f"Server error: {str(e)}")
                print(f"❌ Error processing application: {e}")
        else:
            self.send_error(404, "Endpoint not found")
    
    def log_message(self, format, *args):
        """Custom log format"""
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] {format % args}")

def run_server():
    """Start the application submission server"""
    server_address = ('', PORT)
    httpd = HTTPServer(server_address, ApplicationHandler)
    
    print("=" * 70)
    print("🏢 PepperTree Townhomes - Application Submission Server")
    print("=" * 70)
    print(f"✅ Server running on port {PORT}")
    print(f"📁 Storing applications in: {os.path.abspath(SUBMISSIONS_DIR)}/")
    print(f"🔗 Endpoint: http://localhost:{PORT}/submit")
    print(f"📋 Ready to receive applications...")
    print("=" * 70)
    print("\nPress Ctrl+C to stop the server\n")
    
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n\n🛑 Server stopped by user")
        httpd.shutdown()

if __name__ == '__main__':
    run_server()

File: submit_application.php
<?php
/**
 * Rental Application Submission Handler
 * Works on any standard web server with PHP support
 * Saves applications as JSON files - No email, no dependencies
 */

// Prevent output buffering issues
ob_start();

// Set headers first
header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

// Handle preflight OPTIONS request
if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    ob_end_clean();
    exit;
}

// Only accept POST requests
if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    http_response_code(405);
    ob_end_clean();
    echo json_encode(['success' => false, 'message' => 'Method not allowed. POST requests only.']);
    exit;
}

// Configuration - No dependencies, just file storage
$SUBMISSIONS_DIR = __DIR__ . '/applications';

// Disable error display, log instead
error_reporting(E_ALL);
ini_set('display_errors', 0);
ini_set('log_errors', 1);

try {
    // Get POST data
    $input = file_get_contents('php://input');
    
    if (empty($input)) {
        throw new Exception('No data received');
    }
    
    $data = json_decode($input, true);
    
    if (json_last_error() !== JSON_ERROR_NONE) {
        throw new Exception('Invalid JSON: ' . json_last_error_msg());
    }
    
    // Add server metadata
    $data['submittedAt'] = date('c'); // ISO 8601 format
    $data['submittedFrom'] = $_SERVER['REMOTE_ADDR'] ?? 'Unknown';
    $data['userAgent'] = $_SERVER['HTTP_USER_AGENT'] ?? 'Unknown';
    $data['serverTime'] = time();
    
    // Create submissions directory if it doesn't exist
    if (!is_dir($SUBMISSIONS_DIR)) {
        if (!mkdir($SUBMISSIONS_DIR, 0755, true)) {
            throw new Exception('Failed to create applications directory');
        }
    }
    
    // Generate unique filename
    $timestamp = date('Ymd_His');
    $firstName = isset($data['firstName']) ? preg_replace('/[^a-zA-Z0-9]/', '_', $data['firstName']) : 'Unknown';
    $lastName = isset($data['lastName']) ? preg_replace('/[^a-zA-Z0-9]/', '_', $data['lastName']) : 'Unknown';
    $filename = $timestamp . '_' . $firstName . '_' . $lastName . '.json';
    $filepath = $SUBMISSIONS_DIR . '/' . $filename;
    
    // Ensure unique filename (in case of simultaneous submissions)
    $counter = 1;
    while (file_exists($filepath)) {
        $filename = $timestamp . '_' . $firstName . '_' . $lastName . '_' . $counter . '.json';
        $filepath = $SUBMISSIONS_DIR . '/' . $filename;
        $counter++;
    }
    
    // Save to JSON file with pretty print
    $json_data = json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    
    if ($json_data === false) {
        throw new Exception('Failed to encode JSON: ' . json_last_error_msg());
    }
    
    $bytes_written = file_put_contents($filepath, $json_data, LOCK_EX);
    
    if ($bytes_written === false) {
        throw new Exception('Failed to write application file');
    }
    
    // Optional: Log to server error log (if configured)
    error_log("Rental application saved: $filename");
    
    // Clean output buffer and send success response
    ob_end_clean();
    http_response_code(200);
    echo json_encode([
        'success' => true,
        'message' => 'Application submitted successfully',
        'filename' => $filename,
        'timestamp' => $data['submittedAt']
    ]);
    
} catch (Exception $e) {
    // Log error to server log
    error_log("Rental application error: " . $e->getMessage());
    
    // Clean output buffer and send error response
    ob_end_clean();
    http_response_code(500);
    echo json_encode([
        'success' => false,
        'message' => 'Error processing application. Please try again or contact us directly.'
    ]);
}
?>
